### ISSUE TITLE
Crash with many objects with geoCoordinates props: "panic: runtime error: makeslice: len out of range"

### ISSUE THREAD

**Issue by klibro** [weaviate/weaviate#2539]
I have had an issue where the weaviate server crashes (and is unable to recover) when I attempt an import of a postcode file that is publicly accessible at https://www.getthedata.com/open-postcode-geo

On my system, their error occurs after around 88,000 records have been processed, but this varies. I have had it work fine once (when my docker volume wasn't mapped) and it has failed twice (when my volume was mapped).

I attach the Jupyter notebook which contains the schema definition and file import code.
[PopulatePostcodes.ipynb.zip](https://github.com/semi-technologies/weaviate/files/10435263/PopulatePostcodes.ipynb.zip)

**Comment by etiennedi** [comment]
Note for person investigating this: As mentioned already during the discussion on Slack, an early suspect here could be the `geoCoordinates` prop. We have a lot of large-scale tests, including manual ones up to billion scale that haven't produced this error. But if we can get it to crash with such a small data set, the one thing that immediately stands out to me is the `geoCoordinates` prop. Maybe it has something to do with it.

**Comment by etiennedi** [comment]
Thanks @klibro for the report and the repro example!

**Comment by klibro** [comment]
Just a few more pertinent points: I am running this on Docker on a MacBook Air with M2 processor, and I have tried it with docker images v1.17 and :latest (id faee0ec67f76bfe895af579b0b93adde10e4a0694ea579574f2a6aeab5fb34bc)

**Comment by klibro** [comment]
Tried it a third time but this time reducing the batch size down from 100 to 50. Got through to 103550 records but then crashed. So the good news is that this is definitely reproducible, at least on my machine.

**Comment by etiennedi** [comment]
That's great news, our team mostly uses M1-based Macs, so I hope our envs won't be too different.

**Comment by etiennedi** [comment]
Good news, everyone!
![image](https://user-images.githubusercontent.com/8974479/212972485-6a450601-00e0-4491-92e5-2d62f799d93d.png)

I just ran your notebook and I'm getting the same weird behavior. It is indeed related to the index for the geo property. Thanks, we'll take it from here and hopefully have a fix ready soon!

**Comment by etiennedi** [comment]
It seems to initally go wrong with the commit log merge on the `geoCoordinates` prop's HNSW index (geo indexes are also backed by HNSW, just like vector indexes):

```
ERRO[0117] hnsw commit log maintenance (combining) failed  action=hnsw_commit_log_combining error="combine \"./data/postcode_G7V6dJtrwKbW_location.hnsw.commitlog.d/1673977027.condensed\" and \"./data/postcode_G7V6dJtrwKbW_location.hnsw.commitlog.d/1673977077.condensed\": rename and clean up files: rename tmp (\"./data/postcode_G7V6dJtrwKbW_location.hnsw.commitlog.d/1673977027.combined.tmp\") to final (\"./data/postcode_G7V6dJtrwKbW_location.hnsw.commitlog.d/1673977027\"): rename ./data/postcode_G7V6dJtrwKbW_location.hnsw.commitlog.d/1673977027.combined.tmp ./data/postcode_G7V6dJtrwKbW_location.hnsw.commitlog.d/1673977027: no such file or directory"
```

Everything else is then probably a follow-up issue.

**Comment by mickey-liu** [comment]
Hi @etiennedi, @klibro, I ran into this issue as well, even when I didn't have a volume mounted to the /var/lib/weaviate dir, this bug is unrelated to volume mounts right?

**Comment by etiennedi** [comment]
Correct, I was able to reproduce it running completely without Docker.

I'll make sure we give this some priority. 7 days open and no fix yet is too long for this kind of bug. Sorry about that.

**Comment by etiennedi** [comment]
Hi @klibro and @mickey-liu 

could you try again with this Docker image: `semitechnologies/weaviate:preview-fix-crash-with-many-geo-props-0f3f418`

This should fix it.

**Comment by mickey-liu** [comment]
@etiennedi looks like that image fixed it! Thanks @etiennedi

**Comment by etiennedi** [comment]
Cool! Thanks for the feedback. It requires a bit more testing and then we can include it in the upcoming `v1.17.2` patch.

**Comment by klibro** [comment]
Thanks. This seems to have fixed it for me as well. Imported all 1.7 million rows whereas previously was crashing out after 90 thousand or so. I had managed to work around the problem with the previous docker images by putting a time.sleep(10) call in my import loop allowing my import to "take a breather" every few thousand records. Does this tally with the fix that you made? Something along the lines of buffers not having enough time to flush or similar?

My bigger concern is not so much with the crash, but with the fact that the database was left in a broken state from which it tried but failed to recover. In a test scenario the solution was to trash the data and retry. If this had been a production system I'd have had a very serious headache. Are there things that can be done when the automatic recovery fails?

Anyway, thanks for the speedy fix.

**Comment by etiennedi** [comment]
I did a bit more digging and I'm starting to understand why we never ran into this with a test pipeline before. I tried to add a new test pipeline (independent of your reproducing example) and the pipeline got to 2M objects without a single crash with v1.17.1. So I started comparing the two and it seems that the open-postcode dataset contains a lot of duplicate geo coordinates. The HNSW index isn't very good at handling duplicates, so it has to do a lot more work. That additional work made the race much more likely to happen.

I changed my test pipeline from 2M unique data points to repeating the same 20 or so data points over and over again and the crash happened in the first minute. 

> My bigger concern is not so much with the crash, but with the fact that the database was left in a broken state from which it tried but failed to recover. 

Yeah, the crash was a side-effect of the file corruption. With the corruption fixed, both issues are gone.

> Are there things that can be done when the automatic recovery fails?

Yes, but in this case, some data loss would have occurred because Weaviate was actively "sabotaging" the state on disk. The objects themselves would have been recoverable, but the index for the location points would have been unrecoverable. So if this had been a prod situation, you could have deleted the index, then weaviate would have started up again, and you could have recovered your data, but would still have had to reimport to re-build the index.

**Comment by etiennedi** [comment]
By the way, future releases will have to go through an automated stress pipeline that prevents this. We already had those in place for "regular" properties. But not for `geoCoordinates` props. That will be added now. Better late than never ;-)

**Comment by etiennedi** [comment]
For those interested, you can see the new stress test pipeline here. Here's an example failiure with `v1.17.1`: https://github.com/weaviate/weaviate-chaos-engineering/actions/runs/4008711822/jobs/6883217761#step:4:159

**Comment by etiennedi** [comment]
Fixed in `v1.17.2`. Seems the GitHub auto-close didn't work. Closing manually.
